zkevm-circuits/src/evm_circuit/execution/error_invalid_opcode.rs

```diff
@@ -0,0 +1,234 @@
+            rw_counter_end_of_reversion.expr(),
+        );
+
+        // Go to EndTx only when is_root.
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
+        );
+
+        // When it's a root call.
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(2.expr() + cb.curr.state.reversible_write_counter.expr()),
+                ..StepStateTransition::any()
+            });
+        });
+
+        // When it is an internal call, need to restore caller's state as finishing this
+        // call. Restore caller state to next StepState.
+        let restore_context = cb.condition(not::expr(cb.curr.state.is_root.expr()), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+            )
+        });
+
+        // Constrain `RwCounterEndOfReversion`.
+        let rw_counter_end_of_step =
+            cb.curr.state.rw_counter.expr() + cb.rw_counter_offset() - 1.expr();
+        cb.require_equal(
+            "rw_counter_end_of_reversion = rw_counter_end_of_step + reversible_counter",
+            rw_counter_end_of_reversion.expr(),
+            rw_counter_end_of_step + cb.curr.state.reversible_write_counter.expr(),
+        );
+
+        Self {
+            opcode,
+            op_lt256,
+            rw_counter_end_of_reversion,
+            restore_context,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = F::from(step.opcode.unwrap().as_u64());
+        self.opcode.assign(region, offset, Value::known(opcode))?;
+        self.op_lt256.assign(region, offset, opcode, 256.into())?;
+
+        self.rw_counter_end_of_reversion.assign(
+            region,
+            offset,
+            Value::known(F::from(call.rw_counter_end_of_reversion as u64)),
+        )?;
+
+        self.restore_context
+            .assign(region, offset, block, call, step, 2)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::test::rand_bytes;
+    use crate::test_util::CircuitTestBuilder;
+    use eth_types::bytecode::Bytecode;
+    use eth_types::{bytecode, ToWord, Word};
+    use itertools::Itertools;
+    use lazy_static::lazy_static;
+    use mock::TestContext;
+
+    const TESTING_CALL_DATA_PAIRS: [(usize, usize); 2] = [(0x20, 0x00), (0x1010, 0xff)];
```

Comment:
I think this won't affect the invalid opcode exceptions right? If we can remove it and simplify the tests.

Reply:
Deleted internal call test cases in 759ed1332d040e5426d364e93377499c20e33e1a. Since no effects for this invalid opcode error.

Reply:
Ah sorry I didn't mean to remove the internal test, I mean we don't need to test different call data pairs. I think it's still useful to test internal call with invalid opcode.

Reply:
Sorry for my misunderstanding. I reverted the previous change, and fixed to only keep on call data offset and length pair (0x20, 0x00) for internal test in commit [b8cbf2e](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1089/commits/b8cbf2eaf3578d7c98c1309f5389050459aa449d).

Reply:
Thanks!